src/main/java/org/mskcc/oncokb/curation/util/CdxUtils.java

```diff
@@ -0,0 +1,230 @@
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.mskcc.oncokb.curation.domain.CompanionDiagnosticDevice;
+import org.mskcc.oncokb.curation.domain.FdaSubmission;
+import org.mskcc.oncokb.curation.domain.FdaSubmissionType;
+import org.mskcc.oncokb.curation.service.FdaSubmissionTypeService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CdxUtils {
+
+    private final Logger log = LoggerFactory.getLogger(CdxUtils.class);
+
+    private final String CDX_URL =
+        "https://www.fda.gov/medical-devices/in-vitro-diagnostics/list-cleared-or-approved-companion-diagnostic-devices-in-vitro-and-imaging-tools";
+
+    private final Set<String> cdxDeviceFieldNames = Set.of("Device Name", "Trade Name", "Device");
+
+    private final Set<String> cdxNumberFieldNames = Set.of("PMA Number", "HDE Number", "510(K) Number", "De Novo Number");
+
+    @Autowired
+    private FdaSubmissionTypeService fdaSubmissionTypeService;
+
+    // Extract relevant information from the FDA CDx page
+    public List<CompanionDiagnosticDevice> getCdxListFromHTML() throws IOException {
+        // Fetch the html from the fda cdx page
+        Document document = Jsoup.connect(CDX_URL).get();
+
+        Element table = document.select("table").first();
+
+        // Extract the table rows
+        List<CompanionDiagnosticDevice> companionDiagnosticDevices = new ArrayList<>();
+        Element tbody = table.select("tbody").first();
+        Integer rowSpanCount = 0;
+        for (Element tr : tbody.getElementsByTag("tr")) {
+            Elements td = tr.getElementsByTag("td");
+            if (rowSpanCount > 0) { // Some cells may span across multiple rows
+                rowSpanCount--;
+                continue;
+            }
+            try {
+                rowSpanCount = Integer.parseInt(td.first().attributes().get("rowspan")) - 1;
+            } catch (NumberFormatException e) {}
+            ArrayList<String> tableCells = td.stream().limit(3).map(e -> e.text()).collect(Collectors.toCollection(ArrayList::new));
+
+            // Create CompanionDiagnosticDevice entity
+            CompanionDiagnosticDevice cdx = new CompanionDiagnosticDevice();
+            if (tableCells.size() < 3) {
+                continue; // Skip table cells with less than 3 elements
```

Comment:
Could you add a log.warn here specifying which line is skipped?

Reply:
Sure.